### PR TITLE
refactor(WebSocketShard): payload sending

### LINKED ISSUE #8918: @discordjs/ws should leave padding in ratelimit queue for heartbeats
### Which package is the feature request for?

ws

### Feature

The ratelimit queue should leave room for 5 heartbeats or similar important payloads (consider a case where a bout in many guilds fetch all members, you'd end up timing out)

alongside that, payloads like heartbeats should be sent whenever they are sent while respecting the ratelimit queue (so if you have a queue of 500 payloads, the heartbeats should be sent as soon as they're available)

### Ideal solution or implementation

The queue should be split into 2: 115 payloads off any kind and 5 important payloads that have priority. This queue should handle the maximum size per minute, as well as the separate buckets

### Alternative solutions or implementations

_No response_

### Other context

_No response_

### LINKED ISSUE #10089: The WebSocket rate limit has been hit, this should never happen
### Which package is this bug report for?

discord.js

### Issue description

I'm getting this error message time to time `The WebSocket rate limit has been hit, this should never happen`,

My app is doing quite a lot of requests but nothing crazy, getting members `await members.fetch()` prob 10req/s max and then add/remove user roles > could ramp up to 100req/s max, prob average less than 1req/s.

Socket restarts automatically afterwards but usually I get `Members didn't arrive in time` or other bunch of issues when it restarts, if you have any ideas on why this happens, would be appreciated thanks.

### Code sample
```
[1661756] 2024-01-14T10:04:23.412Z debug: [WS => Shard 0] The WebSocket rate limit has been hit, this should never happen
[1661756] 2024-01-14T10:04:23.412Z debug: [WS => Shard 0] Destroying shard
        Reason: none
        Code: 4008
        Recover: Reconnect
[1661756] 2024-01-14T10:04:23.412Z debug: [WS => Shard 0] Connection status during destroy
        Needs closing: false
        Ready state: 3
[1661756] 2024-01-14T10:04:23.912Z debug: [WS => Shard 0] Connecting to wss://gateway.discord.gg?v=10&encoding=json
[1661756] 2024-01-14T10:04:23.918Z debug: [WS => Shard 0] Waiting for event hello for 60000ms
[1661756] 2024-01-14T10:04:24.082Z debug: [WS => Shard 0] Preparing first heartbeat of the connection with a jitter of 0.036720235270546686; waiting 1514ms
[1661756] 2024-01-14T10:04:24.083Z debug: [WS => Shard 0] Waiting for identify throttle
[1661756] 2024-01-14T10:04:24.083Z debug: [WS => Shard 0] Identifying
        shard id: 0
        shard count: 1
        intents: 98819
        compression: none
[1661756] 2024-01-14T10:04:24.083Z debug: [WS => Shard 0] Waiting for event ready for 15000ms
[1661756] 2024-01-14T10:04:24.172Z debug: [WS => Shard 0] Tried to send a non-crucial payload before the shard was ready, waiting
[1661756] 2024-01-14T10:04:25.602Z debug: [WS => Shard 0] First heartbeat sent, starting to beat every 41250ms
[1661756] 2024-01-14T10:04:26.537Z debug: [WS => Shard 0] Heartbeat acknowledged, latency of 935ms.
[1661756] 2024-01-14T10:04:26.687Z debug: Failed to find guild, or unknown type for channel 118265xxx 16
[1661756] 2024-01-14T10:04:26.688Z debug: Failed to find guild, or unknown type for channel 118264xxx 16
[1661756] 2024-01-14T10:04:26.804Z debug: [WS => Shard 0] Shard received all its guilds. Marking as fully ready.
```

### Versions

"discord.js": "14.10.2",
node 20.10.0
typescript 4.8.2

### Issue priority

Low (slightly annoying)

### Which partials do you have configured?

Not applicable

### Which gateway intents are you subscribing to?

Guilds, GuildMembers

### I have tested this issue on a development release

_No response_

### PR DESCRIPTION
**Pull request by didinele**
**Please describe the changes this PR makes and why it should be merged:**
After consulting with Discord and some conversations in an internal thread[^1], this is the resolution we came up with for the following:

- Fixes #8918
- Fixes #10089

[^1]: https://discord.com/channels/222078108977594368/1196094076395716608/1196094076395716608

### CODE REVIEW

File: packages/ws/src/ws/WebSocketShard.ts [hunk 1]
```diff
@@ -369,34 +380,36 @@ export class WebSocketShard extends AsyncEventEmitter<WebSocketShardEventsMap> {
 
 		await this.sendQueue.wait();
 
-		if (--this.sendRateLimitState.remaining <= 0) {
-			const now = Date.now();
-
-			if (this.sendRateLimitState.resetAt > now) {
-				const sleepFor = this.sendRateLimitState.resetAt - now;
+		const now = Date.now();
+		if (now > this.sendRateLimitState.resetAt) {
```
Comment (vladfrangu) [inline packages/ws/src/ws/WebSocketShard.ts:384]: Should we maybe make this `>=`?

**Review by vladfrangu — APPROVED** [review]

**Review by Jiralite — CHANGES_REQUESTED** [review]

File: packages/ws/src/ws/WebSocketShard.ts [hunk 2]
```diff
@@ -75,8 +75,8 @@ export enum CloseCodes {
 }
 
 export interface SendRateLimitState {
-	remaining: number;
```
[the remark this answers is not in the document]
Comment (didinele) [reply packages/ws/src/ws/WebSocketShard.ts:78]: yup! just ignore it. it's one of those cases were like, maybe that function and type shouldn't even really be exported and you cannot possibly imagine a world where anyone is using those for a good reason in their code.

it's a price I'm willing to pay

[the remark this answers is not in the document]
Comment (RedGuy12) [reply packages/ws/src/ws/WebSocketShard.ts:78]: should it be marked deprecated then?

[the remark this answers is not in the document]
Comment (didinele) [reply packages/ws/src/ws/WebSocketShard.ts:78]: no, what good is deprecating a property that is already being removed... I'm just saying that it's so un-important and that if you use this you're doing something so deeply weird I'm willing to compromise "sneaking" a technically breaking change into a minor/patch release.

[the remark this answers is not in the document]
Comment (RedGuy12) [reply packages/ws/src/ws/WebSocketShard.ts:78]: oh sorry, i meant deprecating the function `getInitialSendRateLimitState()`

**Review by SpaceEEC — CHANGES_REQUESTED** [review]
Very important, I know.

**Review by SpaceEEC — APPROVED** [review]

**Review by kyranet — APPROVED** [review]

**Pull request merged**